README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# SDLfilter

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3631115.svg)](https://doi.org/10.5281/zenodo.3631115)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/SDLfilter)](https://cran.r-project.org/package=SDLfilter)

## Overview

SDLfilter contains a variety of functions to screen GPS/Argos locations
and to assess the adequacy of sample size of tracking data for animal
distribution analysis.

## Installation

``` r
# The official version from CRAN:
install.packages("SDLfilter")

# Or the development version from GitHub:
install.packages("devtools")
devtools::install_github("TakahiroShimada/SDLfilter")
```

## Usage

``` r
library(SDLfilter)
```

### 1. Location filtering

There are three main filtering functions.

-   *dupfilter* filters temporal and spatial duplicates.  
-   *ddfilter* filters locations with high error.  
-   *depthfilter* filters locations by water depth.

<!-- <p>&nbsp;</p> -->

#### 1-1. Load tracking data

``` r
data(turtle)
```

#### 1-2. Remove temporal and spatial duplicates

``` r
turtle.dup <- dupfilter(turtle)
```

#### 1-3. Remove biologically unrealistic fixes

``` r
## Calculate the maximum linear speed between two consecutive locations
V <- vmax(turtle.dup)  

## Calculate the maximum one-way linear speed of a loop trip
VLP <- vmaxlp(turtle.dup) 

## Run ddfilter
turtle.dd <- ddfilter(turtle.dup, vmax=V, vmaxlp=VLP)
```

#### 1-4. Plot data

<details>
<summary>
Click to show code
</summary>

``` r
 # Entire area
 p1 <- to_map(turtle.dup, bgmap=Australia, point.size = 2, line.size = 0.5, axes.lab.size = 0, 
             sb.distance=200, multiplot = FALSE, point.bg = "red",
             title.size=15, title="Entire area")[[1]] + 
   geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, fill="yellow", shape=21)+
   geom_point(aes(x=x, y=y), data=data.frame(x=c(154, 154), y=c(-22, -22.5)), 
              size=3, fill=c("yellow", "red"), shape=21) + 
   annotate("text", x=c(154.3, 154.3), y=c(-22, -22.5), label=c("Retained", "Removed"), 
            colour="black", size=4, hjust = 0)

 # Zoomed in
 p2 <- to_map(turtle.dup, bgmap=SandyStrait, xlim=c(152.7, 153.2), ylim=(c(-25.75, -25.24)), 
             axes.lab.size = 0, sb.distance=10, point.size = 2, point.bg = "red", line.size = 0.5, 
             multiplot = FALSE, title.size=15, title="Zoomed in")[[1]] + 
 geom_path(aes(x=lon, y=lat), data=turtle.dd, size=0.5, colour="black", linetype=1) + 
 geom_point(aes(x=lon, y=lat), data=turtle.dd, size=2, colour="black", shape=21, fill="yellow")
 
 ## plot
 gridExtra::grid.arrange(p1, p2, layout_matrix=cbind(1,2))
```

</details>

![](man/figures/unnamed-chunk-7-1.png)

### 2. Assessing sample sizes (probability-based approach)

#### 2-1. Input UDs

The input data can be either a matrix or a list of RasterLayer objects.
Each row of the matrix or each RasterLayer object contains the
probability distribution of an animal. The function assumes that each
column of a matrix is associated with a unique geographical location,
therefore it is critical that the grid size and geographical extent are
consistent across UDs. In this example (15 tracks of flatback turtles),
the grid size was 1km and the geographical extent was 1901789, 1972789,
-2750915, -2653915 (EPSG:3577) across all 15 layers.

<details>
<summary>
Click to show an example code of UD estimation.
</summary>

``` r
library(adehabitatHR); library(raster)

## Tracking data
data(flatback)
flatback <- track_param(flatback, param = "time") # calculate time between successive locations
flatback_id <- unique(flatback$id)


## Data range with 5km buffer
buff <- 5e+3
xmin <- min(flatback$x) - buff; xmax <- max(flatback$x) + buff
ymin <- min(flatback$y) - buff; ymax <- max(flatback$y) + buff

## Make a grid layer
cell.size <- 1e+3 # (1km x 1km)
x <- seq(xmin, xmax, cell.size)
y <- seq(ymin, ymax, cell.size)
xy.df <- expand.grid(x = x, y = y)
xy.coords <- SpatialPixels(SpatialPoints(xy.df))
xy.sp <- SpatialPoints(xy.coords, proj4string = CRS("+init=epsg:3577"))
z <- rep(1, nrow(xy.df))
xyz <- cbind(xy.df, z)
grid_spdf <- SpatialPixelsDataFrame(xy.coords, xyz)

## UD per turtle
ud_raster <- list()
for(i in 1:length(flatback_id)){
  
  ## ID
  ID <- flatback_id[i]

  ## Tracking data
  turtle.data <- with(flatback, flatback[id %in% ID, ])

  ## Creates an object of class Itraj
  data.ltraj <- with(turtle.data, as.ltraj(turtle.data[,c("x", "y")], date=DateTime, id=ID, burst=ID))
  
  ## Parameters for BRB
  TM = quantile(turtle.data$pTime, probs = 0.95, na.rm = TRUE)*3600
  LM = 50
  dp = BRB.likD(data.ltraj, Tmax=TM, Lmin=LM)
  HM = 100
  
  ## Estimate the UD
  ud <- BRB(data.ltraj, D=dp, Tmax=TM, Lmin=LM, hmin=HM, grid=grid_spdf, type="UD")
  
  ## Convert the UD to raster
  ud_raster[[i]] <- raster(ud)
}
```

</details>

``` r
# A matrix
data(ud_matrix)

# Or a list of RasterLayer/SpatRaster
data(ud_raster)
```

#### 2-2. Calculate overlap probability from 2000 random permutation (\> sample size x 100)

It will take some time to run this code depending on the number of
iterations and the machine specs. The runtime was about 2.5 minutes for
2000 iterations on a linux machine (Intel i7-4702HQ CPU @ 2.20GHz, 16GB
RAM).

``` r
overlap <- boot_overlap(ud_matrix, R = 2000, method = "PHR")
```

#### 2-3. Find the minimum sample size required to estimate the general distribution

An asymptote could be considered once the mean overlap probability
exceeded 95% of the estimated horizontal asymptote (Shimada et al 2021).
The sample size linked to this value is deemed to be the minimum sample
size required to represent the general distribution of the group.

``` r
a <- asymptote(overlap, upper.degree = 5, estimator = 'glm', family = binomial)
```

#### 2-4. Plot the estimated overlap probabilities relative to the sample sizes (black points), the 95% confidence intervals (grey lines), and 95% of the estimated asymptote (dashed line).

<details>
<summary>
Click to show code
</summary>

``` r
ggplot(data = a$results, aes(x = x))+
  geom_errorbar(aes(ymin = y_lwr, ymax = y_upr), width = 0.2, colour = 'darkgrey', size = 1) + 
  geom_point(aes(y = y), size = 2) + 
  geom_hline(yintercept = a$h.asymptote*0.95, linetype = 2) +
  scale_x_continuous(breaks = seq(0, 15, 3), limits = c(1.9,15.1), name = "Animals tracked (n)") +
  scale_y_continuous(limits = c(0.5,1), name = "Overlap probability") +
  theme_light()
```

</details>

![](man/figures/unnamed-chunk-13-1.png)

> Please see the package help pages and Shimada et al. (2012, 2016,
> 2021) for more details.

## References

If you use the function *ddfilter*, please cite

Shimada T, Jones R, Limpus C, Hamann M (2012) Improving data retention
and home range estimates by data-driven screening. *Mar Ecol Prog Ser*
457:171-180 doi:
[10.3354/meps09747](http://dx.doi.org/10.3354/meps09747)

If you use the functions *dupfilter* or *depthfilter*, please cite

Shimada T, Limpus C, Jones R, Hazel J, Groom R, Hamann M (2016) Sea
turtles return home after intentional displacement from coastal foraging
areas. *Mar Biol* 163:1-14 doi:
[10.1007/s00227-015-2771-0](http://dx.doi.org/10.1007/s00227-015-2771-0)

If you use the functions *asymptote*, *boot_overlap*, *combn_overlap* or
*boot_area*, please cite

Shimada, T, Thums, M, Hamann, M, et al. (2021) Optimising sample sizes
for animal distribution analysis using tracking data. *Methods Ecol
Evol* 12(2):288-297 doi:
[10.1111/2041-210X.13506](https://doi.org/10.1111/2041-210X.13506)

## Current version

Version 2.3.2 (2023-04-17)


NEWS
Version 2.3.2 (2023-04-17)

  o fixed bugs in the ddfilter_speed when "method = 2" is selected.
  o significantly improved the processing speed for the ddfilter_speed, especially when "method = 1" is used.


Version 2.3.1 (2023-01-16)

  o added an additional argument "ia" in the distfilter. "ia" is the inner angle between consecutive locations that can be used for detecting potential outliers. Default (NA) is set to ignore this argument.
  o added an additional argument "nloc" in the vmaxlp function. This allows to specify the minimum number of successive locations to be considered a loop trip. The default value is set to be 5 locations.
  o fixed an error in the vmaxlp function that occurs when there is not enough data for the calculation.
  o fixed bugs in the "ML" method for the vmax and vmaxlp functions.
  

Version 2.3.0 (2022-10-11)

  o improved the estimation of vmax of vmaxlp by implementing Maximum Likelihood estimation.
  o fixed minor bugs in the ddfilter, dupfilter and track_param functions.
  o reinstated the kml_track function as the dependent package (plotKML) has been reinstated to the CRAN repository.
  o added a new function to_kmz, which generates KMZ from locations and track.
  

Version 2.2.1 (2022-05-15)

  o added a parallel computing option to the dupfilter_time, dupfilter_space, and dupfilter functions.
  o removed the kml_track function as the dependent package (plotKML) has been removed from the CRAN repository.
  o updated to use the sf and stars packages instead of the sp and raster packages. The updated functions are boot_area, boot_overlap, combn_overlap, depthfilter, dupfilter_space, track_param, dupfilter_time, and map_track
  o updated to use the dplyr package instead of the plyr package. The updated functions are boot_area, ddfilter_loop, ddfilter_speed, dupfilter_qi, dupfilter_space, dupfilter_time, and track_param.  
  o updated the track_param function to use the geosphere package instead of the trip package.
  o The map_track function has been replaced with the "to_map" function.


Version 2.2.0 (2022-01-10)

  o added an option in the asymptote function to estimate the confidence intervals for bootstrapped overlap estimates. The new default is to use the CIs when considering an asymptote.
  o added a new function 'combn_overlap' as an alternative to 'boot_overlap'. This new function calculates overlaps between all possible combination of UDs relative to sample size.
  o improved the processing speed of 'boot_overlap'.


Version 2.1.2 (2021-10-31)

  o fixed a minor bug related with the max.asymptote argument within the asymptote function.


Version 2.1.1 (2021-07-04)

  o amended the ddfilter, dupfilter, and their depending functions (e.g. ddfilter_speed) to return movement parameters used in each function.


Version 2.1.0 (2021-04-02)

  o improved the asymptote function.
  o added a new citation for functions boot_overlap, boot_area and asymptote. Shimada T, Thums M, Hamann M, Limpus CJ, Hays GC, FitzSimmons NN, Wildermann NE, Duarte CD, Meekan MG (2021) Optimising sample sizes for animal distribution analysis using tracking data. Methods in Ecology and Evolution, 12(2):288-297. doi:10.1111/2041-210X.13506 
  o updated the README file.
  o improved the R documentations of functions asymptote, boot_overlap, and boot_area.
  o renamed the example data "curtis" to "flatback".


Version 2.0.1 (2020-07-04)

  o fixed bugs in the track_param function. 
  o The input qi values can be now either the number of GPS satellites or Argos Location Classes. Argos Location Classes will be converted to numerical values, where "A", "B", "Z" will be replaced with "-1", "-2", "-3" respectively.
  o The input DateTime (GMT) can be now either in class POSIXct or character with the following format "2012-06-03 01:33:46".
  o improved the R documentations.


Version 2.0.0 (2020-05-04)

  o added four new functions (boot_overlap, boot_area, asymptote, percent_vol) to assess sample sizes of animal tracking data.
  o added a new function (track_param) to calculate some parameters of animal tracking data (i.e. time, distance, speed, angle between locations).
  o added a new function (kml_track) to visualise the tracking data.
  o added a new example data "curtis".
  o added a new argument "type" in the depthfilter function.
  o renamed some functions (ddfilter.loop -> ddfilter_loop; ddfilter.speed -> ddfilter_speed; dupfilter.exact -> dupfilter_exact; dupfilter.qi -> dupfilter_qi; dupfilter.space -> dupfilter_space; dupfilter.time -> dupfilter_time, distantfilter -> distfilter).
  o renamed an argument in ddfilter (maxvlp -> vmaxlp).
  o renamed the plotMap to map_track
  o improved the map_track function.
  o improved the R documentations.
  o improved the processing speed of the dupfilter function and its dependent functions.
  o updated the vmax and maxvlp functions. In the previous version, when input data contain duplicate locations, the function ceased with error messages. These errors are now avoided by removing duplicate locations using the dupfilter function prior to calculation of vmax and maxvlp.


Version 1.2.1 (2019-02-19)

  o amended "plotMap" in accordance with the ggsn update to version 0.5.0. 


Version 1.2.0 (2019-02-18)

  o added a mew function "distance_filter". This function removes fixes located beyond a given threshold distance.
  o fixed bugs in the example codes.
  o amended "plotMap" in accordance with the ggmap update to version 3.0.0. 


Version 1.1.1 (2018-06-10)


R/depthfilter.r
#' @aliases depthfilter
#' @title Filter locations by water depth
#' @description Function to remove fixes located at a given height from the high tide line.
#' @param sdata A data frame containing columns with the following headers: "id", "DateTime", "lat", "lon", "qi". 
#' This filter is independently applied to a subset of data grouped by the unique "id". 
#' "DateTime" is date & time in class POSIXct. "lat" and "lon" are the recorded latitude and longitude in decimal degrees. 
#' "qi" is the numerical quality index associated with each fix where the greater number represents better quality 
#' (e.g. number of GPS satellites used for estimation).
#' @param bathymetry object of class "RasterLayer" containing bathymetric data in meters. Geographic coordinate system is WGS84.
#' @param extract Method to extract cell values from raster layer inherited from extract function of raster package. 
#' Default is "bilinear". See the raster package for details.
#' @param tide A data frame containing columns with the following headers: "tideDT", "reading", "standard.port". 
#' "tideDT" is date & time in class POSIXct for each observed tidal height. "reading" is the observed tidal height in meters. 
#' "standard.port" is the identifier of each tidal station.
#' @param qi An integer specifying threshold quality index. 
#' Fixes associated to a quality index higher than the threshold are excluded from the depthfilter. Default is 4
#' @param depth An integer denoting vertical difference from a high tide line in meters. 
#' A positive value indicates above the high tide and a negative value indicates below the high tide. 
#' The function removes fixes above the given threshold. Default is 0 m (i.e. high tide line).
#' @param tidal.plane A data frame containing columns with the following headers: 
#' "standard.port", "secondary.port", "lat", "lon", "timeDiff", "datumDiff". 
#' "standard.port" is the identifier for a tidal observation station. 
#' "secondary.port" is the identifier for a station at which tide is only predicted using tidal records observed at the related standard port. 
#' "lat" and "lon" are the latitude and longitude of each secondary port in decimal degrees. 
#' "timeDiff" is the time difference between standard port and its associated secondary port. 
#' "datumDiff" is the baseline difference in meters between bathymetry and tidal observations/predictions 
#' if each data uses different datum (e.g. LAT and MSL). 
#' @param filter Default is TRUE. If FALSE, the function does not filter locations but the depth estimates are returned.
#' @import sp raster
#' @importFrom data.table data.table
#' @export
#' @details This function removes fixes located at a given height from estimated high tide line when the "filter" option is enabled. 
#' The function chooses the closest match between each fix and tidal observations or predictions in temporal and spatial scales 
#' in order to estimate height of high tide at the time and location of each fix. 
#' It does not filter data when the "filter" option is disabled but it returns the estimated water depth of each location with 
#' the tide effect accounted for (bathymetry + tide height). The estimated water depths are returned in the "depth.exp" column. 
#' @return Input data is returned with two columns added; "depth.exp", "depth.HT". 
#' "depth.exp" is the estimated water depth at the time of location fixing. 
#' "depth.HT" is the estimated water depth at the high tide nearest to the time and location of each fix. 
#' When the "filter" option is enabled, the fixes identified by this filter are removed from the input data. 
#' @author Takahiro Shimada
#' @note Input data must not contain temporal or spatial duplicates.
#' @references Shimada T, Limpus C, Jones R, Hazel J, Groom R, Hamann M (2016) 
#' Sea turtles return home after intentional displacement from coastal foraging areas. 
#' Marine Biology 163:1-14 doi:10.1007/s00227-015-2771-0
#' @references Beaman, R.J. (2010) Project 3DGBR: A high-resolution depth model for the Great Barrier Reef and Coral Sea. 
#' Marine and Tropical Sciences Research Facility (MTSRF) Project 2.5i.1a Final Report, MTSRF, Cairns, Australia, pp. 13 plus Appendix 1.
#' @seealso dupfilter, ddfilter
#' @examples
#' 
#' ### Load data sets
#' # Fastloc GPS data obtained from a green turtle
#' data(turtle)
#' 
#' # Bathymetry model developed by Beaman (2010)
#' data(bathymodel)
#' 
#' # A tidal plane for the example site
#' data(tidalplane)
#' 
#' # Tidal observations and predictions for the example site
#' data(tidedata)
#' 
#' # Map for the example site
#' data(basemap)
#' 
#' 
#' ### Remove temporal and/or spatial duplicates
#' turtle.dup <- dupfilter(turtle)
#' 
#' 
#' ### Remove biologically unrealistic fixes 
#' ## Estimate vmax
#' vmax <- est.vmax(turtle.dup)
#' 
#' ## Estimate maxvlp
#' maxvlp <- est.maxvlp(turtle.dup)
#' 
#' ## Apply ddfilter
#' turtle.dd <- ddfilter(turtle.dup, vmax=vmax, maxvlp=maxvlp)
#'
#'
#' ### Apply depthfilter
#' turtle.dep <- depthfilter(sdata=turtle.dd, 
#'                           bathymetry=bathymodel, 
#'                           tide=tidedata, 
#'                           tidal.plane=tidalplane)
#' 
#' 
#' ### Plot data on a map before and after depthfilter is applied
#' par(mfrow=c(1,2))
#' par(mar=c(4,5,2,1))
#' plot(basemap, col="grey", xlim=c(152.8, 153.1), ylim=c(-25.75, -25.24))
#' axis(1, at=seq(from=152, to=154, by=0.2))
#' axis(2, at=seq(from=-26, to=-25, by=0.2), las=2)
#' mtext("Longitude", side=1, line=2.5)
#' mtext("Latitude", side=2, line=3.5)
#' box()
#' title("Not applied")
#' LatLong <- data.frame(Y=turtle.dup$lat, X=turtle.dup$lon)
#' coordinates(LatLong) <- ~X+Y
#' proj4string(LatLong) <- CRS("+proj=longlat +ellps=WGS84 +datum=WGS84")
#' plot(LatLong, pch=21, bg="yellow", add=TRUE)
#' 
#' par(mar=c(4,4,2,2))
#' plot(basemap, col="grey", xlim=c(152.8, 153.1), ylim=c(-25.75, -25.24))
#' axis(1, at=seq(from=152, to=154, by=0.2))
#' axis(2, at=seq(from=-26, to=-25, by=0.2), las=2)
#' mtext("Longitude", side=1, line=2.5)
#' box()
#' title("Applied")
#' LatLong <- data.frame(Y=turtle.dep$lat, X=turtle.dep$lon)
#' coordinates(LatLong) <- ~X+Y
#' proj4string(LatLong) <- CRS("+proj=longlat +ellps=WGS84 +datum=WGS84")
#' plot(LatLong, pch=21, bg="yellow", add=TRUE)



depthfilter<-function(sdata, bathymetry, extract="bilinear", tide, qi=4, depth=0, tidal.plane, filter=TRUE) {
  
  OriginalSS<-nrow(sdata)
  
  ### Sort data in alphabetical and chronological order
  # Animal data
  sdata<-with(sdata, sdata[order(id, DateTime),])
  row.names(sdata)<-1:nrow(sdata)
  
  # tidal data
  tide<-with(tide, tide[order(standard.port, tideDT),])
  row.names(tide)<-1:nrow(tide)
  
  # tidal plane
  tidal.plane<-with(tidal.plane, tidal.plane[order(standard.port, secondary.port),])
  row.names(tidal.plane)<-1:nrow(tidal.plane) 
  
  
  ### Set lat and lon as "SpatialPoints"
  # Animal data
  LatLong<-data.frame(Y=sdata$lat, X=sdata$lon)
  coordinates(LatLong)<-~X+Y
  proj4string(LatLong)<-CRS("+proj=longlat +ellps=WGS84 +datum=WGS84")
  
  # tidal data
  LatLong.tide<-data.frame(Y=tidal.plane$lat, X=tidal.plane$lon)
  coordinates(LatLong.tide)<-~X+Y
  proj4string(LatLong.tide)<-CRS("+proj=longlat +ellps=WGS84 +datum=WGS84")
  
  
  ### extract bathymetry at each animal location
  sdata$bathy<-extract(bathymetry, LatLong, method=extract)
  
  
  ### Find the nearest tidal station for each animal location  
  # Get nearest tidal ports
  distance.to.ports<-pointDistance(LatLong, LatLong.tide, lonlat=T)
  if(nrow(tidal.plane)==1){
    np.index<-1 
  } else {
    np.index<-apply(distance.to.ports, 1, which.min)    
  }
  
  sdata$nearest.port<-tidal.plane[np.index, "secondary.port"]
  np.list<-levels(factor(sdata$nearest.port))
  
  
  # Get the names of standard ports associated to the nearest ports
  s.index<-match(sdata$nearest.port, tidal.plane$secondary.port)
  sdata$standard.port<-tidal.plane[s.index, "standard.port"]
  
  
  ### Subset tidal data according to the range of animal data
  sp.list<-levels(factor(sdata$standard.port))
  tide<-tide[tide$standard.port %in% sp.list,]
    
  trimTide<-function(j){
      MaxInterval<-max(diff(tide[tide$standard.port %in% j, "tideDT"]))
      units(MaxInterval)<-"mins"
      minDateTime<-min(sdata[sdata$standard.port %in% j, "DateTime"])-MaxInterval
      maxDateTime<-max(sdata[sdata$standard.port %in% j, "DateTime"])+MaxInterval
      with(tide, tide[tideDT>=minDateTime & tideDT<=maxDateTime & standard.port %in% j,])
  }
  
  tide.list<-lapply(sp.list, trimTide)
  tide<-do.call(rbind, lapply(tide.list, data.frame, stringsAsFactors=FALSE))
  
  
  ### Organize tidal data
  # Calculate difference in tide to a subsequent record
  tide$range<-unlist(tapply(tide$reading, tide$standard.port, function(x) c(diff(x),NA)))   
  
  
  # Calculate tidal height increment per minute
  Interval<-unlist(tapply(tide$tideDT, tide$standard.port, function(x) c(as.numeric(diff(x)),NA)))
  tide$increment<-tide$range/Interval
  
  
  
  ## Estimate tide at secondary ports
  GetTideSP<-function(j){
      pri.port<-with(tidal.plane, tidal.plane[secondary.port %in% j, "standard.port"])
      sec.port<-with(tidal.plane, tidal.plane[secondary.port %in% j, "secondary.port"])
      tide.sub<-tide[tide$standard.port %in% pri.port,]
      
      timeDiff <- with(tidal.plane[tidal.plane$secondary.port %in% sec.port,], timeDiff)
      tide.sub$tideDT <- tide.sub$tideDT+timeDiff*60
      tide.sub$secondary.port<-sec.port
      tide.sub
  }
      
  tideS.list<-lapply(np.list, GetTideSP)
  tide.s<-do.call(rbind, lapply(tideS.list, data.frame, stringsAsFactors=FALSE))
  
  
  
  #### Estimate water depth with the tide accounted for
  ### Actual water depth
  # Estimate tidal height at each location 
  tide.s$tideDT1<-tide.s$tideDT
  sdata$DateTime1<-sdata$DateTime
  
  DateTime<-NULL; increment<-NULL; reading<-NULL; rm2<-NULL; rm3<-NULL; tideDT<-NULL
  
  tidalHeight<-function(j){
      ## Make data tables
      tide.dt<-data.table(tide.s[tide.s$secondary.port %in% j,], key="tideDT1")
      sdata.dt<-data.table(sdata[sdata$nearest.port %in% j,], key="DateTime1")
      
      ## Get the differences in time between each location and tidal reading at secondary ports
      estTide<-tide.dt[sdata.dt, list(DateTime, tideDT, reading, increment), roll="nearest"]
      estTide$gap.time<-with(estTide, as.numeric(difftime(tideDT, DateTime, units = "mins")))
      
      ## Estimate tidal adjustment for each location
      sdata.dt$adj.reading<-with(estTide, reading+increment*gap.time)
      
      as.data.frame(sdata.dt)
  }
  
      
  sdataDT.list<-lapply(np.list, tidalHeight)
  sdata<-do.call(rbind, lapply(sdataDT.list, data.frame, stringsAsFactors=FALSE))
  
  
  # Estimate actual depth of each location with tide effect
  d.index<-match(sdata$nearest.port, tidal.plane$secondary.port)
  adj.datum<-tidal.plane[d.index, "datumDiff"]
  sdata$depth.exp<-with(sdata, bathy+adj.datum-adj.reading)
  
  
  if(filter %in% TRUE){
      ### Water depth at closest high tide
      ## Organize tidal data
      #Estimate high tide at each port: (0=High, 1=others)
      peakTide<-function(j){
          tide.temp<-tide.s[tide.s$secondary.port %in% j,]
          Interval<-as.numeric(with(tide.temp, difftime(tide.temp[2, "tideDT"], tide.temp[1, "tideDT"], units="mins")))
          ncell.lookup<-120/Interval
          
          find.peakHL<-function(i) {
              if(tide.temp$reading[i]>=max(tide.temp$reading[(i+1):(i+ncell.lookup)]) & tide.temp$reading[i]>=max(tide.temp$reading[(i-1):(i-ncell.lookup)])){
                  0
              } else {
                  1
              }
          }
          
          peak<-unlist(lapply(ncell.lookup:(nrow(tide.temp)-ncell.lookup-1), find.peakHL))
          tide.temp$peak<-c(rep(NA, ncell.lookup-1), peak, rep(NA, ncell.lookup+1))
          tide.temp[tide.temp$peak == 0,]
      }
      
      tide.s<-with(tide.s, tide.s[order(secondary.port, tideDT),])
      row.names(tide.s)<-1:nrow(tide.s)
      
      peakTide.list<-lapply(np.list, peakTide)
      Htide<-do.call(rbind, lapply(peakTide.list, data.frame, stringsAsFactors=FALSE))
      Htide<-Htide[!(is.na(Htide$reading)),]
      
      
      ## Estimate height of high tide at each turtle location
      HtideHeight<-function(j){
          Htide.dt<-data.table(Htide[Htide$secondary.port %in% j,], key="tideDT1")
          sdata.dt<-data.table(sdata[sdata$nearest.port %in% j,], key="DateTime1")
          estHtide<-Htide.dt[sdata.dt, list(DateTime, tideDT, reading, increment), roll="nearest"]
          estHtide$adj.reading<-with(estHtide, reading)  
          sdata.dt$adj.reading.HT<-estHtide$adj.reading
          
          as.data.frame(sdata.dt)
      }
      
      sdataHTH.list<-lapply(np.list, HtideHeight)
      sdata<-do.call(rbind, lapply(sdataHTH.list, data.frame, stringsAsFactors=FALSE))
      
      
      ## Estimate actual depth of a location at the nearest high tide
      d.index<-match(sdata$nearest.port, tidal.plane$secondary.port)
      adj.datum<-tidal.plane[d.index, "datumDiff"]
      sdata$depth.HT<-with(sdata, bathy+adj.datum-adj.reading.HT)
      
      
      ### Remove locations according to water depth at high tide: (0 = remove, 1 = keep)
      sdata<-with(sdata, sdata[!(qi<=qi & depth.HT>depth),])
      
      
      #### Re-order data
      sdata<-with(sdata, sdata[complete.cases(lat),])
      sdata<-with(sdata, sdata[order(id, DateTime),])
      row.names(sdata)<-1:nrow(sdata)
      
      
      #### Report the summary of filtering
      FilteredSS<-nrow(sdata)
      RemovedSamplesN<-OriginalSS-FilteredSS
      RemovedSamplesP<-round((1-(FilteredSS/OriginalSS))*100,2)
      DepthEst<-nrow(sdata[!(is.na(sdata$depth.exp)),])
      
      cat("\n")
      cat("Input data:", OriginalSS, "locations")
      cat("\n")
      cat("Filtered data:", FilteredSS, "locations")
      cat("\n")
      cat("depthfilter removed ", RemovedSamplesN, " locations (", RemovedSamplesP, "% of original data)", sep="")
      cat("\n")
      cat("Actual water depth (bathymetry + tide height) was estiamted for", DepthEst, "of", OriginalSS, "locations")
      cat("\n\n")
      
      
      #### Delete working columns and return the output
      drops<-c("nearest.port", "bathy", "standard.port", "adj.reading", "adj.reading.HT", "DateTime1")
      sdata<-sdata[,!(names(sdata) %in% drops)] 
      return(sdata)
      
  } else {
      
      #### Re-order data
      sdata<-with(sdata, sdata[order(id, DateTime),])
      row.names(sdata)<-1:nrow(sdata)
      
      
      #### Report the summary of filtering
      DepthEst<-nrow(sdata[!(is.na(sdata$depth.exp)),])

      cat("\n")
      cat("No location was removed by depthfilter (i.e. filter option was desabled)")
      cat("\n")
      cat("Actual water depth (bathymetry + tide height) was estiamted for", DepthEst, "of", OriginalSS, "locations")
      cat("\n\n")
      
      
      #### Delete working columns and return the output
      drops<-c("nearest.port", "bathy", "standard.port", "adj.reading", "DateTime1")
      sdata<-sdata[,!(names(sdata) %in% drops)] 
      return(sdata)
         
  }
}
